File: src/main/java/com/binarysushi/studio/language/isml/codeInsight/tags/ISMLTagDescriptor.java
package com.binarysushi.studio.language.isml.codeInsight.tags;

import com.intellij.html.impl.RelaxedHtmlFromSchemaElementDescriptor;
import com.intellij.openapi.util.Condition;
import com.intellij.psi.PsiElement;
import com.intellij.psi.impl.source.html.dtd.HtmlNSDescriptorImpl;
import com.intellij.psi.impl.source.xml.XmlDocumentImpl;
import com.intellij.psi.util.PsiTreeUtil;
import com.intellij.psi.xml.XmlAttribute;
import com.intellij.psi.xml.XmlTag;
import com.intellij.util.ArrayUtil;
import com.intellij.util.containers.ContainerUtil;
import com.intellij.xml.XmlAttributeDescriptor;
import com.intellij.xml.XmlElementDescriptor;
import com.intellij.xml.XmlElementsGroup;
import com.intellij.xml.XmlNSDescriptor;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.Nullable;

import java.util.HashMap;
import java.util.Map;

public class ISMLTagDescriptor implements XmlElementDescriptor {
    protected final String name;
    protected final PsiElement declaration;
    private final Map<String, String> attrMap = new HashMap<String, String>();

    public ISMLTagDescriptor(String myName, PsiElement declaration) {
        this.name = myName;
        this.declaration = declaration;

        attrMap.put("isactivedatacontext", "current_category");
        attrMap.put("iscache", "status,type,hour,minute,varyby");
        attrMap.put("iscomponent", "pipeline,locale");
        attrMap.put("iscontent", "type,charset,encoding,compact");
        attrMap.put("iscookie", "name,value,comment,domain,path,maxAge,version,secure");
        attrMap.put("isdecorate", "template");
        attrMap.put("isif", "condition");
        attrMap.put("iselseif", "condition");
        attrMap.put("isinclude", "template,url");
        attrMap.put("isloop", "items,var,alias,status,begin,end,step");
        attrMap.put("ismodule", "template,name");
        attrMap.put("isobject", "object,view");
        attrMap.put("isprint", "value,style,formatter,timezone,padding,encoding");
        attrMap.put("isredirect", "location,permanent");
        attrMap.put("isremove", "name,scope");
        attrMap.put("isselect", "name,iterator,description,value,condition,encoding");
        attrMap.put("isset", "name,value,scope");
        attrMap.put("isslot", "id,context,context-object,description,preview-url");
        attrMap.put("isstatus", "value");
    }

    @Override
    public String getQualifiedName() {
        return name;
    }

    @Override
    public String getDefaultName() {
        return name;
    }

    @Override
    public XmlElementDescriptor[] getElementsDescriptors(XmlTag context) {
        XmlDocumentImpl xmlDocument = PsiTreeUtil.getParentOfType(context, XmlDocumentImpl.class);
        if (xmlDocument == null) return EMPTY_ARRAY;
        return xmlDocument.getRootTagNSDescriptor().getRootElementsDescriptors(xmlDocument);
    }

    @Override
    public XmlElementDescriptor getElementDescriptor(XmlTag childTag, XmlTag contextTag) {
        XmlTag parent = contextTag.getParentTag();
        if (parent == null) return null;
        final XmlNSDescriptor descriptor = parent.getNSDescriptor(childTag.getNamespace(), true);
        return descriptor == null ? null : descriptor.getElementDescriptor(childTag);
    }

    @Override
    public XmlAttributeDescriptor[] getAttributesDescriptors(@Nullable XmlTag context) {
        if (context != null) {
            final String tagName = context.getName();
            if (attrMap.containsKey(tagName)) {
                final String[] attrs = attrMap.get(tagName).split(",");
                final XmlAttributeDescriptor[] result = new XmlAttributeDescriptor[attrs.length];
                for (int i = 0; i < attrs.length; i++) {
                    result[i] = new ISMLXmlAttributeDescriptor(tagName, attrs[i]);
                }
                return result;
            }
        }
        final XmlAttributeDescriptor[] commonAttributes = HtmlNSDescriptorImpl.getCommonAttributeDescriptors(context);
        return RelaxedHtmlFromSchemaElementDescriptor.addAttrDescriptorsForFacelets(context, commonAttributes);
    }

    @Nullable
    @Override
    public XmlAttributeDescriptor getAttributeDescriptor(XmlAttribute attribute) {
        return getAttributeDescriptor(attribute.getName(), attribute.getParent());
    }

    @Nullable
    @Override
    public XmlAttributeDescriptor getAttributeDescriptor(@NonNls final String attributeName, @Nullable XmlTag context) {
        return ContainerUtil.find(getAttributesDescriptors(context), new Condition<XmlAttributeDescriptor>() {
            @Override
            public boolean value(XmlAttributeDescriptor descriptor) {
                return attributeName.equals(descriptor.getName());
            }
        });
    }

    @Override
    public XmlNSDescriptor getNSDescriptor() {
        return null;
    }

    @Nullable
    @Override
    public XmlElementsGroup getTopGroup() {
        return null;
    }

    @Override
    public int getContentType() {
        return CONTENT_TYPE_ANY;
    }

    @Nullable
    @Override
    public String getDefaultValue() {
        return null;
    }

    @Override
    public PsiElement getDeclaration() {
        return declaration;
    }

    @Override
    public String getName(PsiElement context) {
        return getName();
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public void init(PsiElement element) {

    }

    @Override
    public Object[] getDependences() {
        return ArrayUtil.EMPTY_OBJECT_ARRAY;
    }
}

File: src/main/kotlin/com/binarysushi/studio/language/isml/codeInsight/tags/ISMLTagDescriptor.kt
package com.binarysushi.studio.language.isml.codeInsight.tags

import com.intellij.html.impl.RelaxedHtmlFromSchemaElementDescriptor
import com.intellij.psi.PsiElement
import com.intellij.psi.impl.source.html.dtd.HtmlNSDescriptorImpl
import com.intellij.psi.impl.source.xml.XmlDocumentImpl
import com.intellij.psi.util.PsiTreeUtil
import com.intellij.psi.xml.XmlAttribute
import com.intellij.psi.xml.XmlTag
import com.intellij.util.containers.ContainerUtil
import com.intellij.xml.XmlAttributeDescriptor
import com.intellij.xml.XmlElementDescriptor
import com.intellij.xml.XmlElementsGroup
import com.intellij.xml.XmlNSDescriptor
import org.jetbrains.annotations.NonNls

class ISMLTagDescriptor(private val name: String, private val declaration: PsiElement) : XmlElementDescriptor {
    private val attrMap: MutableMap<String, String> = HashMap()
    override fun getQualifiedName(): String {
        return name
    }

    override fun getDefaultName(): String {
        return name
    }

    override fun getElementsDescriptors(context: XmlTag): Array<XmlElementDescriptor> {
        val xmlDocument = PsiTreeUtil.getParentOfType(
            context,
            XmlDocumentImpl::class.java
        )
            ?: return XmlElementDescriptor.EMPTY_ARRAY
        return xmlDocument.rootTagNSDescriptor.getRootElementsDescriptors(xmlDocument)
    }

    override fun getElementDescriptor(childTag: XmlTag, contextTag: XmlTag): XmlElementDescriptor? {
        val parent = contextTag.parentTag ?: return null
        val descriptor = parent.getNSDescriptor(childTag.namespace, true)
        return descriptor?.getElementDescriptor(childTag)
    }

    override fun getAttributesDescriptors(context: XmlTag?): Array<XmlAttributeDescriptor?> {
        if (context != null) {
            val tagName = context.name
            if (attrMap.containsKey(tagName)) {
                val attrs = attrMap[tagName]!!.split(",").toTypedArray()
                val result = arrayOfNulls<XmlAttributeDescriptor>(attrs.size)
                for (i in attrs.indices) {
                    result[i] = ISMLXmlAttributeDescriptor(tagName, attrs[i])
                }
                return result
            }
        }
        val commonAttributes = HtmlNSDescriptorImpl.getCommonAttributeDescriptors(context)
        return RelaxedHtmlFromSchemaElementDescriptor.addAttrDescriptorsForFacelets(context, commonAttributes)
    }

    override fun getAttributeDescriptor(attribute: XmlAttribute): XmlAttributeDescriptor? {
        return getAttributeDescriptor(attribute.name, attribute.parent)
    }

    override fun getAttributeDescriptor(attributeName: @NonNls String?, context: XmlTag?): XmlAttributeDescriptor? {
        return ContainerUtil.find(
            getAttributesDescriptors(context)
        ) { descriptor: XmlAttributeDescriptor? -> attributeName == descriptor?.name }
    }

    override fun getNSDescriptor(): XmlNSDescriptor? {
        return null
    }

    override fun getTopGroup(): XmlElementsGroup? {
        return null
    }

    override fun getContentType(): Int {
        return XmlElementDescriptor.CONTENT_TYPE_ANY
    }

    override fun getDefaultValue(): String? {
        return null
    }

    override fun getDeclaration(): PsiElement {
        return declaration
    }

    override fun getName(context: PsiElement): String {
        return getName()
    }

    override fun getName(): String {
        return name
    }

    override fun init(element: PsiElement) {}

    init {
        attrMap["isactivedatacontext"] = "current_category"
        attrMap["iscache"] = "status,type,hour,minute,varyby"
        attrMap["iscomponent"] = "pipeline,locale"
        attrMap["iscontent"] = "type,charset,encoding,compact"
        attrMap["iscookie"] = "name,value,comment,domain,path,maxAge,version,secure"
        attrMap["isdecorate"] = "template"
        attrMap["isif"] = "condition"
        attrMap["iselseif"] = "condition"
        attrMap["isinclude"] = "template,url"
        attrMap["isloop"] = "items,var,alias,status,begin,end,step"
        attrMap["ismodule"] = "template,name"
        attrMap["isobject"] = "object,view"
        attrMap["isprint"] = "value,style,formatter,timezone,padding,encoding"
        attrMap["isredirect"] = "location,permanent"
        attrMap["isremove"] = "name,scope"
        attrMap["isselect"] = "name,iterator,description,value,condition,encoding"
        attrMap["isset"] = "name,value,scope"
        attrMap["isslot"] = "id,context,context-object,description,preview-url"
        attrMap["isstatus"] = "value"
    }
}